vespalib/src/vespa/vespalib/util/spin_lock.h

```diff
@@ -0,0 +1,33 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <atomic>
+
+namespace vespalib {
+
+/**
+ * A spin-lock implementation that favors uncontended performance.
+ * Some measures are taken to reduce the impact of threads waiting to
+ * get the lock since this will not affect the fast-path of obtaining
+ * the lock immediately.
+ *
+ * This implementation satisfies the BasicLockable requirements,
+ * making it work with things like std::lock_guard.
+ **/
+class SpinLock {
+private:
+    std::atomic<bool> _lock;
```

Comment:
Could consider using [`std::atomic_flag`](https://en.cppreference.com/w/cpp/atomic/atomic_flag) here. It has a simpler interface than `std::atomic` but is the only atomic type guaranteed to be lock free (that means the `static_assert` below could be removed). The inner relaxed load loop seems to require a C++20 addition for it, so hopefully we have it..!

Reply:
I switched from atomic_flag to atomic_bool because I want to be able to read the variable without writing it to avoid totally messing up the cache. I do not think gcc is C++20 ready just yet, at least I do not think we have enabled it. It is highly unlikely that we want to run Vespa on a system that does not support atomic bool anyway.

Reply:
Yeah, strangely being able to do [`test(std::memory_order_relaxed)`](https://en.cppreference.com/w/cpp/atomic/atomic_flag/test) is very recent. Reading variables is cutting edge technology